Global Spacebar Play

regulus79 opened the pull request:
This PR refactors how the spacebar is handled by the editors and the `MainWindow` to allow it to play/stop (`Shift+Space` for play/pause) the last played editor no matter if it is in focus or not.

## Changes
- `Editor::togglePlayStop` made public to allow `MainWindow` to call it.
- Instead of a shortcut, `Editor::keyPressEvent` was added which checks if the spacebar is pressed and plays/stops the editor.
- The last used `PlayMode` is now stored in `Song::m_lastPlayMode` and is updated whenever the play mode changes, but not when it changes to `PlayMode::None`.
- Added a new case to `MainWindow::keyPressEvent` to handle the spacebar and call `togglePlayStop` on the right editor.
- `ke->ignore()` added to the end of `PianoRoll::keyPressEvent`, since without it, there were problems with the key event being lost and not triggering the piano roll to play.

## New Changes
- Made `PatternEditor` use Strong Focus like the rest of the editors. This fixes a bug where the pattern editor sometimes would not play even when it was in focus.
- Added `Shift+Space` to play/pause the last played editor.
- Fixed a bug where pressing space after clicking within the mixer would open a dialog to change the volume of the selected channel.
  - Moved the code which handled the spacebar from `MixerView` to `MixerChannelView`, and did some minor refactoring (converting the event filter into just `keyPressEvent`)
  - Also added `e->ignore()` to the default case of `MixerView::keyPressEvent` in order to let the key event be passed up to MainWindow.
  - ~~Added a call to `setFocus` in `MixerChannelView::mousePressEvent`~~ Set `MixerChannelView`'s focus policy to `Qt::StrongFocus` so it will process still the spacebar key event if you click on it.
  - Set the "Add Effect" button's focus policy to `Qt::NoFocus` to prevent the spacebar from triggering it. I feel like this kind of solution is not ideal, and will probably have to be reworked when we decide to implement better keyboard navigation.

~~I did not implement a global `Shift+Space` which appears to pause/play the editors. Do you want me to?~~ Done.

regulus79 commented:
Has anyone experienced issues in this PR with the pattern editor? I recall when implementing it that when I opened an old project, the pattern editor wouldn't play via the space bar until you actually press the play button.

rubiefawn commented on src/gui/editors/Editor.cpp:
```diff
@@ -155,6 +154,16 @@ void Editor::closeEvent(QCloseEvent * event)
 	event->ignore();
  }
 
+ void Editor::keyPressEvent(QKeyEvent * ke)
```
> For pointers or references, always use a single space between the type and '*' or '&', but no space between the '*' or '&' and the variable name

Source: [Qt Coding Style](https://wiki.qt.io/Qt_Coding_Style#Spaces_within_code) since it is not explicitly overidden in the LMMS coding conventions

```suggestion
 void Editor::keyPressEvent(QKeyEvent *ke)
```

rubiefawn commented on include/Editor.h:
```diff
@@ -57,6 +57,11 @@ class Editor : public QMainWindow
 	DropToolBar * addDropToolBar(QWidget * parent, Qt::ToolBarArea whereToAdd, QString const & windowTitle);
 
 	void closeEvent(QCloseEvent * event) override;
+	void keyPressEvent(QKeyEvent * ke) override;
```
> For pointers or references, always use a single space between the type and '*' or '&', but no space between the '*' or '&' and the variable name

Source: [Qt Coding Style](https://wiki.qt.io/Qt_Coding_Style#Spaces_within_code) since it is not explicitly overidden in the LMMS coding conventions

```suggestion
	void keyPressEvent(QKeyEvent *ke) override;
```

rubiefawn reviewed (approved):
I left a few style suggestions, which should be considered non-blocking.

AW1534 reviewed (approved):
Tested this and it works perfectly as described. The code looks good too.

tresf commented:
> ke->ignore() added to the end of PianoRoll::keyPressEvent, since without it, there were problems with the key event being lost and not triggering the piano roll to play.

If possible, I'd like this tested against the qt6 branch as well.  I had to add an [extra `ignore()`](https://github.com/LMMS/lmms/pull/7339/commits/10632b31ddc022639a7b941e5edc539dc63529de) for mouse propagation recently and this PR reminded me of that.
* #7339

tresf commented:
This has some interesting behavior.  This is best shown through a video.  Steps:
1. Load a project
2. Click the title of the song-editor window
3. Hit <kbd>Space</kbd>
4. Click the title of the pattern-editor window
5. Hit <kbd>Space</kbd>
6. Click an empty space in the song-editor window
7. Hit <kbd>Space</kbd>

(Clicking the pattern editor button in the toolbar does not reproduce this behavior)

There are also some nuanced focus issues that are likely unrelated to this PR, such as hitting spacebar clicks a previously focused button.  We may decide to remove certain item focuses when window focus changes, as a separate enhancement.

![pattern_behavior](https://github.com/user-attachments/assets/c26c174b-13bb-489b-9104-1205f198d8ea)

regulus79 commented:
> This has some interesting behavior. This is best shown through a video. Steps:

I can also reproduce this issue. It appears to only occur when loading saved projects. I'm not sure why this is happening, so I will have to investigate further.

regulus79 commented:
For some reason after opening a project, the key event is not passed to the pattern editor even when the window is in focus. This does not occur for the Song Editor or the Piano Roll. It also does not occur the first time you open LMMS. However, the issue resolves itself when you interact with the Pattern Editor by either clicking somewhere in the window background.

regulus79 commented:
Alright, I think I've fixed it! It appears that `PatternEditor` did not share the same `setFocusPolicy` as the other editors, which somehow caused the issue. I don't really understand how Qt's focus system works, but adding those lines seemed to fix the problem.

I also added a `ke->ignore()` to the default branch of `AutomationEditor::keyPressEvent` so that the event will be passed on to `Editor::keyPressEvent` if it is not processed by the pattern editor. This fixed an issue caused by this PR which I was not aware of, which broke the space bar in the automation editor.

regulus79 commented:
> If possible, I'd like this tested against the qt6 branch as well. I had to add an [extra ignore()](https://github.com/LMMS/lmms/pull/7339/commits/10632b31ddc022639a7b941e5edc539dc63529de) for mouse propagation recently and this PR reminded me of that.

I tested out that branch with this pr merged into it, and it seemed to work fine. I did encounter a segfault when adding an effect, but I can't reproduce it and I don't know if it's related.

tresf commented:
> Alright, I think I've fixed it! It appears that PatternEditor did not share the same setFocusPolicy as the other editors, which somehow caused the issue. I don't really understand how Qt's focus system works, but adding those lines seemed to fix the problem.

@regulus79 appears fixed here as well, thanks!  This feature is awesome!

This may be unrelated to this PR, but the mixer doesn't work at all (well, I got it to work once but I don't know how)... instead it always tries to change a knob value.  I assume this is due to the channel gaining focus, but the only way to reliably remove focus is to open an effect and hit space there.  I'm not sure if you'd rather track that issue here or in a separate issue.

What's interesting about this behavior is that it continues to happen even after hiding the mixer.  If you hit space, after hiding the mixer, it tries to adjust the mixer volume still. 😆

regulus79 commented:
Before we merge, are you all sure you don't want me to add a global `Shift+Space` to pause?

sakertooth commented:
> Before we merge, are you all sure you don't want me to add a global Shift+Space to pause?

I would think the global spacebar would work as a toggle.

regulus79 commented:
I'm asking because that's how it works right now. Normal space does play/stop, shift+space does play/pause.

sakertooth commented:
> I'm asking because that's how it works right now. Normal space does play/stop, shift+space does play/pause.

Oh wow I wasn't aware lol, I was thinking the global spacebar was play/pause. I think it would be nice if the global spacebar would align with what was chosen in the song editor (i.e., the |<<, <<, and >> options). We should also add that button for the pattern editor too I think.

sakertooth commented:
Are we already doing this? Sorry I haven't tested the PR yet.

regulus79 commented:
> I think it would be nice if the global spacebar would align with what was chosen in the song editor (i.e., the |<<, <<, and >> options). We should also add that button for the pattern editor too I think.

That is already handled by the editor; this PR just calls the play/stop function. As for the pattern editor, perhaps that would fit in #7794?

sakertooth commented:
Okay, so I suppose adding the `Shift+Space` would be okay for consistency mainly. I'm not really against the addition.

regulus79 commented:
Done!

sakertooth commented on include/Editor.h:
```diff
@@ -57,6 +57,13 @@ class Editor : public QMainWindow
 	void closeEvent(QCloseEvent * event) override;
+	void keyPressEvent(QKeyEvent *ke) override;
+public slots:
+	/// Called by pressing the space key. Plays or stops.
```
`//!` for documentation comments.

```suggestion
	//! Called by pressing the space key. Plays or stops.
```

sakertooth commented on include/Editor.h:
```diff
@@ -57,6 +57,13 @@ class Editor : public QMainWindow
+public slots:
+	/// Called by pressing the space key. Plays or stops.
+	void togglePlayStop();
+	/// Called by pressing shift+space. Toggles pause state.
```
```suggestion
	//! Called by pressing shift+space. Toggles pause state.
```

sakertooth commented on src/gui/MainWindow.cpp:
```diff
@@ -1287,6 +1287,31 @@ void MainWindow::keyPressEvent( QKeyEvent * _ke )
+		case Qt::Key_Space:
+		{
+			Editor* lastEditor = nullptr;
+			switch(Engine::getSong()->lastPlayMode())
```
```suggestion
			switch (Engine::getSong()->lastPlayMode())
```

sakertooth commented on include/Song.h:
```diff
@@ -257,6 +257,7 @@ class LMMS_EXPORT Song : public TrackContainer
 		return m_playMode;
 	}
 
+	inline PlayMode lastPlayMode() const { return m_lastPlayMode; }
```
`inline` is redundant here.
```suggestion
	PlayMode lastPlayMode() const { return m_lastPlayMode; }
```

sakertooth reviewed (commented):
I would make some small formatting adjustments but I won't try and nitpick too much. I'll test the PR and after that we should be good.

Also, I think maybe the PR could use a better title? Something that describes what the commit actually does. Maybe like "Add a global spacebar".

sakertooth commented:
When I press spacebar in the Mixer, it shows this:

![image](https://github.com/user-attachments/assets/b359c756-9d7c-40ab-8378-24bd372198fa)

regulus79 commented:
> When I press spacebar in the Mixer, it shows this:
That also occurs in master. For some reason when you press spacebar after clickin within the mixer, it thinks you are trying to set the volume of the currently selected channel.

sakertooth commented:
Any chance we could fix that in here?

regulus79 commented:
I mean, I guess I could. But I'm not sure it's really related to this PR.

sakertooth commented:
I would think anyone using this feature, this would be somewhat of the first thing they notice. I get it's already on master but I was hoping for this feature to really be useful with the mixer in particular. If its an easy change, I think it would be fine to do it here. Otherwise, if it requires a decent amount of changes from what the PR has right now, then it can wait I guess.

sakertooth commented:
I would also say its perfectly in scope to be honest: The feature is not really useful in the mixer, which is not something to scoff at.

regulus79 commented:
Alright, it should be fixed!

I had to remove the handling of the space key event from `MixerView` and add it to `keyPressEvent` in `MixerChannelView` so that it would only be triggered if the user has the channel in focus. Also, it appeared that `MixerChannelView` was using an `eventFilter` to handle key press events, so I converted that to be `keyPressEvent` to be consistent.

regulus79 commented:
Wait wait wait hold on a second I did something wrong

regulus79 commented:
Yeah, so just like PianoRoll and AutomationEditor, the MixerView's `keyPressEvent` function did not set `ke->ignore()` by default if it did nothing, which meant the key event was not getting passed up to `MainWindow`. It should be fixed now.

sakertooth commented:
Great, this is really handy now! :)

regulus79 commented:
Wait a minute again... somehow I broke the spacebar to enter the volume. I should really be doing more testing before doing `git push`.

Unless we're okay with losing that functionality? If we want spacebar to always mean play/stop/pause, then maybe it would make sense to remove this shortcut/change it to something else?

sakertooth commented:
> Unless we're okay with losing that functionality? If we want spacebar to always mean play/stop/pause, then maybe it would make sense to remove this shortcut/change it to something else?

I wasn't even initially aware that spacebar in the Mixer opened the volume dialog. IMO, since we can already open the dialog by double clicking, opening it with the spacebar might be redundant. But yeah that's just my opinion. It's a hard decision and it needs more input from other devs/the community.

regulus79 commented:
It's okay, I fixed it (for realsies this time :D)

Basically the `MixerChannelView` was not automatically getting focus when you click on it (except when renaming the channel, since in that case the lineEdit child had focus), so I added a call to `setFocus` in `MixerChannelView::mousePressEvent` which seemed to do the trick. I hope I tested it well enough this time.

tresf commented:
> > Unless we're okay with losing that functionality? If we want spacebar to always mean play/stop/pause, then maybe it would make sense to remove this shortcut/change it to something else?
> 
> I wasn't even initially aware that spacebar in the Mixer opened the volume dialog. IMO, since we can already open the dialog by double clicking, opening it with the spacebar might be redundant. But yeah that's just my opinion. It's a hard decision and it needs more input from other devs/the community.

I thought about this too... It would likely become a future accessibility/robots issue.  I'm sure we have plenty of these, but that's the first thing that comes to mind especially when we're talking about automating the GUI as a futuremap.

tresf commented on src/gui/MainWindow.cpp:
```diff
@@ -1287,6 +1287,28 @@ void MainWindow::keyPressEvent( QKeyEvent * _ke )
+				default:
+					getGUI()->songEditor()->togglePlayStop();
+					break;
+			}
```
Completely wrong to recommend this.  Reopening for discussion.  No, no no no no. 😆

tresf replied:
Poll created: https://discord.com/channels/203559236729438208/207159523616489473/1353744413310058588

tresf replied:
Results are in.  Rule is submitted. https://github.com/LMMS/lmms/wiki/Coding-Conventions#switch-statements

rubiefawn replied:
It looks like all the `switch`/`case`s in this file match the updated coding conventions, marking this as resolved

bratpeki commented:
Here are my findings, where the ticked boxes are well implemented, and the unchecked ones would need some discussion.

- [x] When value-input boxes are focused, the song doesn't play
- [x] The space bar works to place a space character in the Note editor
- [x] Before, hitting the space bar after clicking the "Add mixer track" button would repeat the action of adding a mixer track. This has been made so the space bar plays the track instead
- [x] Playing works the Song editor is minimized/hidden

---

- [ ] Hitting the space bar while you've selected something in the browser plays the song only if the Song editor is opened. If it's closed, it doesn't play.
- [ ] Hitting the space bar while a mixer track is selected doesn't play the song and instead opens a value-input box for the mixer track volume

---

I'm perfectly fine with these last two. If @regulus79 considers any of them bugs, please patch accordingly. If not, this is merge-ready to me! Will test some more, just to make sure, but this seems to be great! :grin:

Spacemagehq commented:
https://github.com/user-attachments/assets/7c9c2520-ee95-4b98-baab-8dec82d70d12

when you hit the side it does the same as last time, the space bar just opens and closes it and will not play it
there is also another bug on Windows when you open up the vst window lmms won't play or stop it but when you close out of the vst window it plays and stops the song.

Spacemagehq commented:
https://github.com/user-attachments/assets/617c76a9-a163-462e-8211-d00b98b70b8c

When you hit the top of the mixer it plays and stops just fine, but when you click the strip/channel itself the rename pops up.

bratpeki commented:
> when you hit the side it does the same as last time, the space bar just opens and closes it and will not play it

Can confirm.

regulus79 commented:
> Hitting the space bar while you've selected something in the browser plays the song only if the Song editor is opened. If it's closed, it doesn't play. Same goes for the instrument selection window.

I can't reproduce this, unless you're talking about how sample files preview when you click on them/press space? But I think that might be intentional.

> Hitting the space bar while a mixer track is selected doesn't play the song and instead opens a value-input box for the mixer track volume

> When you hit the top of the mixer it plays and stops just fine, but when you click the strip/channel itself the rename pops up.

This appears to be an intentional feature, since there was code in `Mixer.cpp` specifically for handling the space bar to change the volume of the current channel. I just fixed it so that it was only triggered if you had the mixer channel in focus. Do we want to remove/change this shortcut?

> when you hit the side it does the same as last time, the space bar just opens and closes it and will not play it

I can also reproduce this.

> there is also another bug on Windows when you open up the vst window lmms won't play or stop it but when you close out of the vst window it plays and stops the song.

Do you mean pressing space while you have the VST window selected? (Also, is your vst embedded or not?) I can reproduce it either way; though it seems like it sometimes kind of works when it's embedded.

Spacemagehq commented:
> > Hitting the space bar while you've selected something in the browser plays the song only if the Song editor is opened. If it's closed, it doesn't play. Same goes for the instrument selection window.
> 
> I can't reproduce this, unless you're talking about how sample files preview when you click on them/press space? But I think that might be intentional.
> 
> > Hitting the space bar while a mixer track is selected doesn't play the song and instead opens a value-input box for the mixer track volume
> 
> > When you hit the top of the mixer it plays and stops just fine, but when you click the strip/channel itself the rename pops up.
> 
> This appears to be an intentional feature, since there was code in `Mixer.cpp` specifically for handling the space bar to change the volume of the current channel. I just fixed it so that it was only triggered if you had the mixer channel in focus. Do we want to remove/change this shortcut?
> 
> > when you hit the side it does the same as last time, the space bar just opens and closes it and will not play it
> 
> I can also reproduce this.
> 
> > there is also another bug on Windows when you open up the vst window lmms won't play or stop it but when you close out of the vst window it plays and stops the song.
>

bratpeki commented:
What? :sob:

Spacemagehq commented:
> > Hitting the space bar while you've selected something in the browser plays the song only if the Song editor is opened. If it's closed, it doesn't play. Same goes for the instrument selection window.
> 
> I can't reproduce this, unless you're talking about how sample files preview when you click on them/press space? But I think that might be intentional.
> 
> > Hitting the space bar while a mixer track is selected doesn't play the song and instead opens a value-input box for the mixer track volume
> 
> > When you hit the top of the mixer it plays and stops just fine, but when you click the strip/channel itself the rename pops up.
> 
> This appears to be an intentional feature, since there was code in `Mixer.cpp` specifically for handling the space bar to change the volume of the current channel. I just fixed it so that it was only triggered if you had the mixer channel in focus. Do we want to remove/change this shortcut?
> 
> > when you hit the side it does the same as last time, the space bar just opens and closes it and will not play it
> 
> I can also reproduce this.
> 
> > there is also another bug on Windows when you open up the vst window lmms won't play or stop it but when you close out of the vst window it plays and stops the song.
> 
> Do you mean pressing space while you have the VST window selected? (Also, is your vst embedded or not?) I can reproduce it either way; though it seems like it sometimes kind of works when it's embedded.

 the vst 2 on windows in my lmms is not embedded.

regulus79 commented:
> That sounds good! Could you point me to that commit? I would like to see how you did it.

Basically in https://github.com/LMMS/lmms/pull/7762/commits/755e31f4134c335d489ac7ffc14b02458ed20566 I moved the code which handles the spacebar key event from `MixerView.cpp` to `MixerChannelView.cpp` so it would only be triggered if the mixer channel were in focus, not the entire mixer. (I also reworked the existing event filter into a normal `keyPressEvent` to make things simpler.)

Then, I realized the mixer did not ignore key events by default (which meant they wouldn't be passed on to `MainWindow` to trigger the global spacebar play), so I added that in https://github.com/LMMS/lmms/pull/7762/commits/c42d9d1bd8c41314b1035d5e2fe10aa262fe3403.

Then, I realized that mixer channels by default didn't acquire focus when you clicked on them, so I added a call to `setFocus` in `mouseClickEvent` in https://github.com/LMMS/lmms/pull/7762/commits/e3253d289e3195d51030b49ff663f454ed7a991b. Maybe this can also be done by setting the focus policy? I'm not an expert in Qt focusing.

regulus79 commented:
> when you hit the side it does the same as last time, the space bar just opens and closes it and will not play it

Fixed in #7829.

bratpeki commented:
> Fixed in https://github.com/LMMS/lmms/pull/7829.

Merging that after this, right?

I'll test this one more time for good measure!

tresf commented:
> > Fixed in #7829.
> 
> Merging that after this, right?

That PR is setup as draft.  I'd recommend cherry-picking it here, close that one out, getting a final approval and merging this  immediately after.

regulus79 commented:
> I'd recommend cherry-picking it here, close that one out, getting a final approval and merging this immediately after.

Actually, I realized that PR does not adequately address the issue, since it turns out there are multiple things in the sidebar which can grab the focus. I think it would be best to merge this as is, and then later work on fixing the focus in these edge cases.

szeli1 commented on src/gui/MixerView.cpp:
```diff
@@ -536,15 +536,8 @@ void MixerView::keyPressEvent(QKeyEvent * e)
 		case Qt::Key_F2:
 			renameChannel(m_currentMixerChannel->channelIndex());
 			break;
-		case Qt::Key_Space:
```
Why was this removed? I'm curious

regulus79 replied:
Previously, MixerView handled the space bar as a shortcut to change the volume of the currently selected mixer channel. However, this caused issues were it would handle the space bar event even if the user did not have a channel in focus (like if they were interacting with the effect rack), so I decided to move the functionality to `MixerChannelView` so that it would only trigger when the channel itself is in focus.

szeli1 replied:
Okay, feel free to close these!

szeli1 commented on src/gui/editors/PatternEditor.cpp:
```diff
@@ -49,6 +49,8 @@ PatternEditor::PatternEditor(PatternStore* ps) :
 {
 	setModel(ps);
+	setFocusPolicy(Qt::StrongFocus);
+	setFocus();
```
Why `setFocus()` in constructor? It will only run once.

szeli1 reviewed (approved):
I did not test this PR.

I saw that `MainWindow::keyPressEvent` was used to ignore piano widget focus. I also used this for shortcuts, but is this really the best way to do things? Anyway, approving!

tresf commented:
> > I'd recommend cherry-picking it here, close that one out, getting a final approval and merging this immediately after.
> 
> Actually, I realized that PR does not adequately address the issue, since it turns out there are multiple things in the sidebar which can grab the focus. I think it would be best to merge this as is, and then later work on fixing the focus in these edge cases.

It looks like some last-minute reviews came in, so I'll hold off on merging until they're answered.

szeli1 commented:
Okay, I approve this PR, I hope can get merged soon!

bratpeki commented:
Does that mean this is testing-ready again?

From a video above, I can see that that "space" from the BB editor plays the full song and not the BB pattern. I think that's pretty bad. I'm completely for any other context playing the full song, but a BB editor should play the BB pattern.

regulus79 commented:
> From a video above, I can see that that "space" from the BB editor plays the full song and not the BB pattern.

Which video? I believe that was one of the bugs I fixed.

tresf commented:
> Does that mean this is testing-ready again?
> 
> From a video above, I can see that that "space" from the BB editor plays the full song and not the BB pattern. I think that's pretty bad. I'm completely for any other context playing the full song, but a BB editor should play the BB pattern.

The follow comments explain that this was fixed, quoting:

> > Alright, I think I've fixed it! It appears that PatternEditor did not share the same setFocusPolicy as the other editors, which somehow caused the issue. I don't really understand how Qt's focus system works, but adding those lines seemed to fix the problem.

> @regulus79 appears fixed here as well, thanks! This feature is awesome!

tresf commented:
> Does that mean this is testing-ready again?

Yes,  let's get the convos resolved and some final testing on this.

bratpeki commented:
So that's fixed, great! Testing ASAP

bratpeki commented:
The PR feels great to use, overall:

- If multiple playable elements are in focus (Piano roll, Automation editor, BB editor, Song editor), the last played one is played, unless another one is focused
- The LADSPA and LMMS plugins don't cause issues
- The instruments don't cause issues

I've only found two things I think are odd:

1. Open/close an effect in the mixer, from the "Controls" button. Now, using the spacebar closes a plugin, opens it back up and only then plays the editor in question. It doesn't even have to be exactly that plugin you were opening and closing.
2. When you first load up a project, hitting space from the mixer just changes the volume of the master track.

The second one is somewhat okay for me, mostly a weird edge case, and we can't ensure the user has any playable element open, because we don't force that. So doing what would normally happen when hitting the spacebar while focused on a mixer track is reasonable.

The first thing, however, is most odd. I'd think hitting the spacebar would just play the last focused playable element.

regulus79 commented:
> Open/close an effect in the mixer, from the "Controls" button. Now, using the spacebar closes a plugin, opens it back up and only then plays the editor in question. It doesn't even have to be exactly that plugin you were opening and closing.

Fixed!

> When you first load up a project, hitting space from the mixer just changes the volume of the master track.

I can't seem to reproduce this. When I start up lmms and press space, the song editor plays. To be fair, I am using a custom template, idk if that would make a difference. But even when making a new project with the empty template, the space bar still initially plays the song editor.

regulus79 commented:
Given this has 4 approvals, I think it is reasonable for me to merge it now.

sakertooth commented:
I think we could've had a function in the `Editor` base class to play the last selected editor,, either the Pattern Editor or the Song Editor, Then, any window that presses space can start playing the last selected editor. `MainWindow` can then do the same.That would solve the "I have to add `ke->ignore` and such everywhere and propagate the events correctly to `MainWindow`) problem.

regulus79 commented:
Ooo I like the idea of having like a static function in `Editor`, that would make things much cleaner!

I've sort of implemented it in #7926
